src/pymor/algorithms/qr.py

```diff
@@ -0,0 +1,152 @@
+    kwargs
+        Keyword arguments specific to the method.
+
+    Returns
+    -------
+    Q
+        The orthonormal |VectorArray| of the same length as `V`.
+    R
+        The upper-triangular/trapezoidal matrix.
+
+    Raises
+    ------
+    RuntimeError
+        If the vectors in `V` are detected to be linearly dependent.
+    """
+    assert method == 'gram_schmidt'
+    len_V = len(V)
+    if check:
+        V_orig = V if copy else V.copy()
+    Q, R = gram_schmidt(V, product=product, return_R=True, atol=0, rtol=0, offset=offset,
+                        check=False, copy=copy, **kwargs)
+    if len(Q) < len_V:
+        raise RuntimeError('Could not construct an orthonormal basis of full length.')
+    if check:
+        _check_qr(V_orig, Q, R, product, offset, check_orth_tol, check_recon_tol)
+    return Q, R
+
+
+@defaults('atol', 'rtol', 'check', 'check_orth_tol', 'check_recon_tol', 'method')
+def rrqr(V, product=None, atol=1e-13, rtol=1e-13, offset=0,
+         check=True, check_orth_tol=1e-3, check_recon_tol=1e-3,
+         copy=True, method='gram_schmidt', **kwargs):
+    r"""Compute a rank-revealing QR (RRQR) decomposition.
+
+    For a |VectorArray| `V` and inner product |Operator| `product`,
+    a rank-revealing QR decomposition of `V` is given by any
+    orthonormal |VectorArray| `Q` and
+    upper-triangular/trapezoidal matrix `R` such that
+    the norm of `V - Q.lincomb(R.T)` is "small" and
+    `len(Q) <= len(V)`.
+
+    Parameters
+    ----------
+    V
+        The |VectorArray| for which to compute the rank-revealing QR decomposition.
+    product
+        The inner product |Operator| w.r.t. which to orthonormalize.
+        If `None`, the Euclidean product is used.
+    atol
+        Absolute tolerance used to detect linearly dependent vectors
+        (exact meaning depends on the `method`).
+    rtol
+        Relative tolerance used to detect linearly dependent vectors
+        (exact meaning depends on the `method`).
+    offset
+        Assume that the first `offset` vectors are already orthonormal.
+        Those vectors will not be changed even when `copy` is `False`.
+    check
+        If `True`, check if the resulting `Q` is orthonormal and if `Q*R` approximates `A`.
+    check_orth_tol
+        Tolerance for the orthogonality check:
+        :math:`\lVert Q^H P Q - I \rVert_F \leqslant \mathtt{orth\_tol}`.
+    check_recon_tol
+        Tolerance for the reconstruction check:
+        :math:`\lVert A - Q R \rVert_F \leqslant \mathtt{recon\_tol} \lVert A \rVert_F`.
+    copy
+        If `True`, create a copy of `A` instead of modifying `A` in-place.
+    method
+        QR decomposition method (currently only `'gram_schmidt'`).
+    kwargs
+        Keyword arguments specific to the method.
+
+    Returns
+    -------
+    Q
+        The orthonormal |VectorArray| such that `len(Q) <= len(V)`.
+    R
+        The upper-triangular/trapezoidal matrix.
+    """
```

Comment:
I'd guess that this can also raise an exception in case of breakdown. Maybe it's good to reserve a special exception for that already now?

Reply:
I'm not sure what you mean. I don't see how `rrqr` could break down.

Reply:
Ok, we can leave it out for now and add it later if one implementation can raise some error.